hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSSchedulerNode.java

```diff
@@ -118,23 +178,56 @@ synchronized FSAppAttempt getReservedAppSchedulable() {
    *
    * @param containers container to mark
    */
-  void addContainersForPreemption(Collection<RMContainer> containers) {
-    containersForPreemption.addAll(containers);
+  void addContainersForPreemption(Collection<RMContainer> containers,
+                                  FSAppAttempt appAttempt) {
+    for(RMContainer container : containers) {
+      containersForPreemption.put(container, appAttempt);
+    }
   }
 
   /**
    * @return set of containers marked for preemption.
    */
   Set<RMContainer> getContainersForPreemption() {
-    return containersForPreemption;
+    return containersForPreemption.keySet();
   }
 
   /**
    * Remove container from the set of containers marked for preemption.
+   * Reserve the preempted resources for the app that requested
+   * the preemption.
    *
    * @param container container to remove
    */
-  void removeContainerForPreemption(RMContainer container) {
+  private synchronized void removeContainerForPreemption(RMContainer container) {
```

Comment:
If we add the entry to the map when we add container, this method doesn't need to be changed.